pgms/vquickmorph3d/vquickmorph3d.man
.ds Vi Vista
.ds Vn 2.0
.TH vquickmorph3d 1Vi "15 Februar 1994" "\*(Vi Version \*(Vn"
.SH NAME
vquickmorph3d \- apply a quick 3D morphological filter to a binary image
.SH SYNOPSIS
\fBvquickmorph3d\fR [\fB-\fIoption\fR ...] \
[\fIinfile\fR] [\fIoutfile\fR]
.SH DESCRIPTION
\fBvquickmorph3d\fP
applies a 3D morphological filter to an image by thresholding a distance transform.
Each input image is filtered to 
produce a corresponding output image.
.LP
This program emulates a "true" morphological filter by thresholding
a 3D distance transform. It can only emulate filtering by a sphere
shaped structuring element. It is much faster than the original
morphological filtering. It produces slightly different results.
.LP
The input images must be binary images of type VBit. The output images
will also be binary. 
.SH COMMAND LINE OPTIONS
\fBvquickmorph3d\fP accepts the following options:
.IP \fB-help\fP 15n
Prints a message describing options.
.IP "\fB-in\fP \fIinfile\fP"
Specifies the Vista data file containing the input images.
.IP "\fB-out\fP \fIoutfile\fP"
Specifies where to write the output images as a Vista data file.
.IP "\fB-radius\fP \fIdiameter\fP"
Specifies the radius of a sphere shaped structuring element.
Default: 6.
.IP "\fB-op\fP \fIop\fP"
Specifies the operation as one of the following:
.RS 17n
.PP
.nf
.ta 8n
\fBdilate\fP    dilation
\fBerode\fP     ersion
\fBopen\fP      opening (erosion followed by dilation)
\fBclose\fP     closing (dilation followed by erosion)
.DT
.fi
.RE
.LP
Input and output files can be specified on the command line or allowed to
default to the standard input and output streams.

.SH LITERATURE
For an excellent tutorial on image morphology, see                            
Haralick, R.M, S.R. Sternberg, and X. Zhuang, "Image analysis using     
mathematical morphology," 
.I IEEE Trans. PAMI, 
vol. PAMI\-9, No. 4, July 1987,   
pp. 532-550.                                                            
Or see                                                                   
Maragos, P. and R. Shaffer,                                             
"Morphological systems for multidimensional Signal Processing",         
.I Proc. IEEE, 
vol. 78, no. 4, April 1990.


.SH "SEE ALSO"
.na
.nh
.BR vbinmorph3d (1Vi),
.BR VImage (3Vi),
.BR Vista (7Vi)
.hy
.ad
.SH AUTHOR
Gabriele Lohmann <lohmann@cns.mpg.de>


debian/changelog
via (2.0.1svn191-0) stable; urgency=low

  *renamed via-bin to via-utils

 -- Lipsia Group (The official Lipsia developer group) <lipsia@cbs.mpg.de>  Tue, 09 Nov 2010 12:17:00 +0400 

via (2.0.0-0) stable; urgency=low

  *This version comes with the new lipsia-2.0 package
   and comprises some minor bug fixes
  *Fixed type error of VLong (is now long)

 -- Lipsia Group (The official Lipsia developer group) <lipsia@cbs.mpg.de>  Tue, 09 Nov 2010 12:17:00 +0400 

via (1.6.0-2) unstable; urgency=low

  * New internal release

 -- Frank Burkhardt <burk@cbs.mpg.de>  Mon, 20 Jul 2009 11:45:41 +0200

via (1.6.0-1) unstable; urgency=low

  * Initial release (Closes: #521328).
  * New Upstream Version. Renamed 'vcat' to 'vcatobj' to not conflict with
    atfs package in Debian.
  * Update Standards-version to 3.8.2.0 -- no changes necessary.
  * Update watchfile for recent change in upstream naming scheme.
  * Add versioned dependency on debhelper (>= 5) upon recommendation by
    lintian.

 -- Michael Hanke <michael.hanke@gmail.com>  Fri, 19 Jun 2009 13:50:52 +0200

via (1.5.2+svn3529-2) UNRELEASED; urgency=low

  * Add alternative build-dependency on atlas3-base-dev to be able to build
    the package on Debian etch.
  * FTP access to the upstream sources needs passive mode. Fixed debian/watch.

 -- Michael Hanke <michael.hanke@gmail.com>  Tue, 14 Apr 2009 02:46:10 +0200

via (1.5.2+svn3529-1) UNRELEASED; urgency=low

  * New Upstream SVN snapshot that relocates the VIA lib headers to their new
    cannonical location.

 -- Michael Hanke <michael.hanke@gmail.com>  Tue, 31 Mar 2009 22:08:01 +0200

via (1.5.2-1) UNRELEASED; urgency=low

  * Initial Debian packaging. This is a complete rewrite of the original
    Debian packaging effort done by the upstream developers.

 -- Michael Hanke <michael.hanke@gmail.com>  Thu, 26 Mar 2009 19:46:09 +0100


pgms/vquickmorph3d/vquickmorph3d.c
/****************************************************************
 *
 * Copyright (C) Max Planck Institute 
 * for Human Cognitive and Brain Sciences, Leipzig
 *
 * Author Gabriele Lohmann, 2004, <lipsia@cbs.mpg.de>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * $Id: vquickmorph3d.c 3177 2008-04-01 14:47:24Z karstenm $
 *
 *****************************************************************/

/*! \brief vquickmorph3d - 3D morphological operators using distance transforms

\par Description
3D morphological operators using distance transforms.

\par Usage

        <code>vquickmorph3d</code>

        \param -in     input image
        \param -out    output image
        \param -op     type of operation (dilate | erode | open | close). Default: dilate
        \param -radius radius of spherical structuring element. Default: 3


\par Examples
<br>
\par Known bugs
none.

\file vquickmorph3d.c
\author G.Lohmann, MPI-CBS, 2004
*/



/* From the Vista library: */
#include <viaio/Vlib.h>
#include <viaio/mu.h>
#include <via.h>
#include <viaio/option.h>

/* From the standard C libaray: */
#include <stdio.h>
#include <stdlib.h>


VDictEntry OPDict[] = {
  { "dilate", 0 },
  { "erode", 1 },
  { "open", 2 },
  { "close", 3 },
  { NULL }
};


int main (int argc, char **argv)
{
  static VDouble radius = 3.0;
  static VLong op = 0;
  static VOptionDescRec options[] = {
    { "op", VLongRepn, 1, &op, VOptionalOpt, OPDict,"type of operation" },
    { "radius", VDoubleRepn, 1, & radius, VOptionalOpt, NULL,
      "radius of structuring sphere" }
  };
  FILE *in_file, *out_file;
  VAttrList list;
  VImage src, dest=NULL;
  VAttrListPosn posn;
  char prg[50];	
  sprintf(prg,"vquickmorph3d V%s", getVersion());
  fprintf (stderr, "%s\n", prg);

  /* Parse command line arguments: */
  VParseFilterCmd (VNumber (options), options, argc, argv, &in_file, &out_file);


  /* Read source image(s): */
  if (! (list = VReadFile (in_file, NULL))) exit (1);
  fclose(in_file);


  /* Perform morphological filtering */
  for (VFirstAttr (list, & posn); VAttrExists (& posn); VNextAttr (& posn)) {
    if (VGetAttrRepn (& posn) != VImageRepn)
      continue;
    VGetAttrValue (& posn, NULL, VImageRepn, & src);

    switch (op) {
    case 0:
      dest = VDTDilate(src,NULL,radius);
      if (dest == NULL) exit (1);
      break;

    case 1:
      dest = VDTErode(src,NULL,radius);
      if (dest == NULL) exit (1);
      break;

    case 2:
      dest = VDTOpen(src,NULL,radius);
      if (dest == NULL) exit (1);
      break;

    case 3:
      dest = VDTClose(src,NULL,radius);
      if (dest == NULL) exit (1);
      break;

    default:
      ;
    }

    VSetAttrValue (& posn, NULL, VImageRepn, dest);
    VDestroyImage (src);
    break;
  }

  /* Write the results to the output file: */
  VHistory(VNumber(options),options,prg,&list,&list);
  if (VWriteFile (out_file, list))
    fprintf (stderr, "%s: done.\n",argv[0]);
  return 0;
}

